### PR TITLE
Use coroutines instead of asynctask in ResetDialogPreferenceFragmentCompat

### LINKED ISSUE #5001: Use coroutines instead of asynctask in ResetDialogPreferenceFragmentCompat
#### Problem description
We have multiple old asynctasks across the project. As you know it's better to use coroutines and asynctask is now even deprecated.
`ResetDialogPreferenceFragmentCompat` contains a small asynctask and replacing it should be a good first step.

### PR DESCRIPTION
**Pull request by dczerwinski**
Closes #5001 

#### What has been done to verify that this works as intended?
Tested manually.

#### Why is this the best possible solution? Were any other approaches considered?
Async tasks are deprected. I replaced them with Corutines.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Should not affect users.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.
#### Before submitting this PR, please make sure you have:
- [X] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [X] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [X] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+import org.odk.collect.android.fragments.dialogs.ResetSettingsResultDialog
+import org.odk.collect.android.injection.DaggerUtils
+import org.odk.collect.android.utilities.ProjectResetter
+import org.odk.collect.android.utilities.ProjectResetter.ResetAction.RESET_PREFERENCES
+import org.odk.collect.androidshared.ui.DialogFragmentUtils.dismissDialog
+import org.odk.collect.androidshared.ui.DialogFragmentUtils.showIfNotShowing
+import timber.log.Timber
+import javax.inject.Inject
+
+class ResetDialogPreferenceFragmentCompat :
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: I like converting classes that you are editing to kotlin but could you do that in a separate commit? It's better to have the fix in one commit so that you can clearly see what was changed.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt (hunk 1 above)
Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: Done. I will remember to split it up into two commits in future.

File: collect_app/src/main/java/org/odk/collect/android/tasks/DeleteInstances.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package org.odk.collect.android.tasks
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.odk.collect.android.instancemanagement.InstanceDeleter
+import org.odk.collect.android.listeners.DeleteInstancesListener
+import timber.log.Timber
+
+class DeleteInstances(
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/tasks/DeleteInstances.kt]: Please create a separate pr for this refactoring. This pr should be just about `ResetDialogPreferenceFragmentCompat`

Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/tasks/DeleteInstances.kt]: Done.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        if (cache.isChecked) {
+            resetActions.add(ProjectResetter.ResetAction.RESET_CACHE)
+        }
+        if (resetActions.isNotEmpty()) {
+            showIfNotShowing(
+                ResetProgressDialog::class.java,
+                (_context as CollectAbstractActivity).supportFragmentManager
+            )
+
+            scope.launch(Dispatchers.IO) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: We have a special module `async` for handling such background tasks. Please take a look at `CoroutineScheduler` and its `immediate` method. I believe that construction should be used here.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    CompoundButton.OnCheckedChangeListener {
+
+    @JvmField
+    @Inject
+    var projectResetter: ProjectResetter? = null
+
+    private val job = Job()
+    private val scope = CoroutineScope(Dispatchers.Main + job)
+
+    private var _preferences: AppCompatCheckBox? = null
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: Ohh I don't like those getters... can we just make those fields nullable without getters and use `!!` when they are called?

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+                ProjectResetter.ResetAction.RESET_FORMS -> if (failedResetActions.contains(action)) {
+                    resultMessage.append(
+                        String.format(
+                            _context.getString(R.string.reset_blank_forms_result),
+                            _context.getString(R.string.error_occured)
+                        )
+                    )
+                } else {
+                    resultMessage.append(
+                        String.format(
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: You can use our `getLocalizedString` then a simple form like this:
`resultMessage.append(_context.getLocalizedString(R.string.reset_blank_forms_result, R.string.success))`
should work. Please refactor other cases in this method too.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    private lateinit var _context: Context
+
+    override fun onAttach(context: Context) {
+        _context = context
+        super.onAttach(context)
+        DaggerUtils.getComponent(context).inject(this)
+    }
+
+    public override fun onBindDialogView(view: View) {
+        _preferences = view.findViewById(R.id.preferences)
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: Would setting listeners immediately work? Something like:
```
preferences = view.findViewById<AppCompatCheckBox?>(R.id.preferences).apply {
            setOnCheckedChangeListener(this@ResetDialogPreferenceFragmentCompat)
        }
```

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+import timber.log.Timber
+import javax.inject.Inject
+
+class ResetDialogPreferenceFragmentCompat :
+    PreferenceDialogFragmentCompat(),
+    CompoundButton.OnCheckedChangeListener {
+
+    @JvmField
+    @Inject
+    var projectResetter: ProjectResetter? = null
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: This should be `lateinit var`

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+import org.odk.collect.androidshared.ui.DialogFragmentUtils.dismissDialog
+import org.odk.collect.androidshared.ui.DialogFragmentUtils.showIfNotShowing
+import timber.log.Timber
+import javax.inject.Inject
+
+class ResetDialogPreferenceFragmentCompat :
+    PreferenceDialogFragmentCompat(),
+    CompoundButton.OnCheckedChangeListener {
+
+    @JvmField
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: I think it's redundant.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt (hunk 3 above)
Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: Yep. I forgot about this class. Corrected.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt (hunk 4 above)
Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: Done.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt (hunk 5 above)
Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: Done.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt (hunk 6 above)
Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: Done.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt (hunk 7 above)
Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: Done.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt (hunk 8 above)
Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: Yes. Deleted.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+    @Inject
+    lateinit var scheduler: Scheduler
+
+    private var preferences: AppCompatCheckBox? = null
+    private var instances: AppCompatCheckBox? = null
+    private var forms: AppCompatCheckBox? = null
+    private var layers: AppCompatCheckBox? = null
+    private var cache: AppCompatCheckBox? = null
+
+    private lateinit var _context: Context
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: I think it doesn't make sense to keep this field you can call `getContext()` / `requireContext()` whenever you need it.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+                )
+            )
+        }
+    }
+
+    private fun getPartiallyTransparentColor(color: Int): Int =
+        Color.argb(150, Color.red(color), Color.green(color), Color.blue(color))
+
+    companion object {
+        fun newInstance(key: String?): ResetDialogPreferenceFragmentCompat {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: key should not be nullable I think.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+                }
+            )
+        }
+    }
+
+    private fun handleResult(resetActions: List<Int>, failedResetActions: List<Int>) {
+        val resultMessage = StringBuilder()
+        for (action in resetActions) {
+            when (action) {
+                RESET_PREFERENCES -> if (failedResetActions.contains(action)) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: This builder is soo verbose. I think we could make it shorter by converting every case like (example for the first one):

```
                RESET_PREFERENCES -> resultMessage.append(
                    requireContext().getLocalizedString(
                        R.string.reset_settings_result,
                        if (failedResetActions.contains(action)) R.string.error_occured else R.string.success
                    )
                )
```

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]
It looks good I mentioned some final things we should fix.

**Comment by grzesiek2010** [comment]
@seadowg 
please take a look at this pull request.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt (hunk 10 above)
Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: Yes, done.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt (hunk 11 above)
Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: Done.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt (hunk 9 above)
Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: We need this because after selecting item to reset in the dialog, we show the progress dialog so that the current dialog is detached from `context `and the `getContext()` method or `requireContext()` will return null in the `handleResult()` method.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: I think we're duplicating the pain points of `AsyncTask` here - we've ended up needing to hold onto state so that the "result" of the async work (`onPostExecute` in `AsyncTask` and the `foreground` block in `Scheduler#immediate`) can access it. If `requireContext()` returns `null` it means that the underlying Activity (the `context` for the `Fragment`) has gone along with the UI, so there is no need for us to display anything. We definitely don't want to be holding our own references to a `context` instance for async work as this can cause nasty memory leaks (Activity objects can be huge as they hold the whole view hierarchy).

The pattern to adopt here (as mentioned in [`STATE.md`](https://github.com/getodk/collect/blob/master/docs/STATE.md)) is to use `Scheduler` (coroutines under the hood) and `LiveData` to avoid needing to hold on to `Context` or views and have the view code "react" (through an `observe` call) to the result (rather than being manipulated from a task/coroutines). 

An example of this would be the use of `BlankFormsListViewModel#syncWithServer` - this fires off work in the background with `Scheduler#immediate` and then sets the resulting data on a `LiveData`. The UI code uses `observe` to react to the returned `LiveData` and we're protected from the Activity being closed etc via the lifecycle awareness features in `observe`.

For `ResetDialogPreferenceFragmentCompat`, we probably want to pull out a method (or create a `ViewModel` to host one) that returns a `LiveData` with the reset and failed actions (set from `foreground`) and the fragment can just call `handleResult` (and hide the progress dialog) from an `observe` observation.

**Review by seadowg — CHANGES_REQUESTED** [review]

**Comment by seadowg** [comment]
Another note here: we want to avoid commit messages that require outside context (like "Apply new review suggestions"). We might need to look back at why a change was made and that wouldn't help us. Even if it's just based on a review suggestion the commit message should describe what actually changed (and hopefully why). It's totally OK to use multiline messages if needed.

File: collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt (hunk 9 above)
Comment (dczerwinski) [reply collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt]: [ResetDialogPreferenceFragmentCompat.kt](https://github.com/getodk/collect/pull/5003/files/caee35315dd499b470d00e9cf1c52aa3246e703a#diff-be7a893a9d6744e30e4526689f3e99b0165cc2c8aecee98478345b92e22b7b0d) is dialog. When user clicks button "Reset" on this dialog, then we are showing progress dialog, running resetting in background and as result of resetting we are showing other dialog. All of this was written in the [ResetDialogPreferenceFragmentCompat.kt](https://github.com/getodk/collect/pull/5003/files/caee35315dd499b470d00e9cf1c52aa3246e703a#diff-be7a893a9d6744e30e4526689f3e99b0165cc2c8aecee98478345b92e22b7b0d)  class. Until now, we needed the `_context` variable because after pressing the "Reset" button, the dialog was dissmised, so that the `context` of dialog was null, and we used the previously saved `_context` to display the reset results. I think that was a very bad solution because like @seadowg mentioned there may be memory leak problems. Even using `LiveData` would not help because we wanted to use a `context` from view that no longer displays. 

I solved this problem by moving the responsibility of displaying the results to the fragment from which the above-mentioned dialog is shown. I also added `viewModel` to this fragment and used `LiveData` as recommended.

**Comment by dczerwinski** [comment]
> Another note here: we want to avoid commit messages that require outside context (like "Apply new review suggestions"). We might need to look back at why a change was made and that wouldn't help us. Even if it's just based on a review suggestion the commit message should describe what actually changed (and hopefully why). It's totally OK to use multiline messages if needed.

Thanks for the clarification. I've already corrected the commit names.

**Comment by seadowg** [comment]
Going to close this for the moment. I think we need to some extra thinking about hiding and showing progress and result dialogs and how that interacts with the view model, and it would be good to add tests around some of this before merging, but I don't think we'll be able to come back and focus on it anytime soon.
